Python/Product/Analysis/DocumentBuffer.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.PythonTools.Parsing;
+
+namespace Microsoft.PythonTools.Analysis {
+    class DocumentBuffer {
+        public int Version { get; private set; } = -1;
+        public StringBuilder Text { get; } = new StringBuilder();
+
+        public void Reset(int version, string content) {
+            Version = version;
+            Text.Clear();
+            Text.Append(content);
+        }
+
+        public void Update(IEnumerable<DocumentChangeSet> changes) {
+            foreach (var change in changes) {
+                Update(change);
+            }
+        }
+
+        public void Update(DocumentChangeSet changes) {
+            if (Version >= 0) {
+                if (changes.FromVersion < Version) {
+                    return;
+                } else if (changes.FromVersion > Version) {
+                    throw new InvalidOperationException("missing prior versions");
+                }
+            }
+
+            int lastStart = int.MaxValue;
+            var lineLoc = SplitLines(Text.ToString()).ToArray();
+
+            foreach (var change in changes.Changes) {
+                if (change.WholeBuffer) {
+                    Text.Clear();
+                    if (!string.IsNullOrEmpty(change.InsertedText)) {
+                        Text.Append(change.InsertedText);
+                    }
+                    continue;
+                }
+
+                int start = NewLineLocation.LocationToIndex(lineLoc, change.ReplacedSpan.Start, Text.Length);
```

Comment:
Are the `NewLineLocation` instances still valid after `Text` is changed during first iteration?

Reply:
Changes have to be ordered in reverse order (according to LSP, but we also check on the next line), so the locations after the change are invalid, but we shouldn't try to use them. They're valid enough for the check to work.

Reply:
Sorry, I don't get it.
Imagine we have string `123456` and two changes: 
- remove range `[1..3]`: `123456 -> 156`
- add text `789` into position `2`: `156 -> 157896`

If they are handled in reverse order, we will get:
- add text `789` into position `2`: `123456 -> 127893456`
- remove range `[1..3]`: `127893456 -> 193456`

Reply:
Also, I think reverse order isn't guaranteed. Changes are reversed in `BufferParser.GetUpdatesForSnapshot`, but in `OutOfProcProjectAnalyzer.UpdateContent` they are simply merged:
```cs
                    changes.AddRange(fileChange.changes.Select(c => new LS.TextDocumentContentChangedEvent {
                        range = new SourceSpan(
                            new SourceLocation(c.startLine, c.startColumn),
                            new SourceLocation(c.endLine, c.endColumn)
                        ),
                        text = c.newText
                    }));
```

Reply:
Reverse order of start location, not reverse order of change. I don't recall if there's a non-overlapping requirement, or if it just works out that VS doesn't send overlapping changes in the same update, but this area is *woefully* underspecified and I don't think we have much of a choice other than assuming that our clients (VS/VSC) behave sensibly.

For VS, the changes are guaranteed to be in increasing start location when we get them, so reversing the order to match LSP is fine.

Reply:
Within each snapshot, which is why we recalculate the line locations between each one.

Reply:
So `changes.ToVersion - changes.FromVersion == 1`?

Reply:
According to the LSP spec, yes. For VS, we expect them to match snapshot version numbers, so the increment is only a fallback and we prefer the explicit value if it's provided. VS Code apparently assumes that updates are synchronous and never dropped.

Reply:
Ok, got it.